sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -439,6 +439,14 @@ object StateStore extends Logging {
   @GuardedBy("loadedProviders")
   private val schemaValidated = new mutable.HashMap[StateStoreProviderId, Option[Throwable]]()
 
+  private val threadPoolLock = new Object
```

Comment:
(No need to address, just thinking out loud)
I wonder there is a way to abstract this better, e.g. put all the details for maintenance task into MaintenanceTask class. But it does not look like be feasible unless we make a change to MaintenanceTask so that it contains the actual task implementation. I'd prefer that, but does not need to be the scope of this PR.

Reply:
Same for new threadPool and partitions - they are two different instances which could be ideally abstracted into one. Though I'm still OK with deferring refactor to later - at least the change follows the way we did, and refactor may deal with two levels of pools altogether.

Reply:
Sure. I think this refactor might be out of the scope of this PR, but this makes sense to me.